infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/RemoveLocalProjectsFolderOnWorkspaceRemove.java

```diff
@@ -0,0 +1,105 @@
+import java.util.concurrent.Executors;
+import org.eclipse.che.api.core.model.workspace.Workspace;
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.core.notification.EventSubscriber;
+import org.eclipse.che.api.workspace.server.event.WorkspaceRemovedEvent;
+import org.eclipse.che.commons.lang.concurrent.LoggingUncaughtExceptionHandler;
+import org.eclipse.che.commons.lang.concurrent.ThreadLocalPropagateContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Listener for remove workspace files after {@code WorkspaceRemovedEvent}.
+ *
+ * @author Alexander Andrienko
+ * @author Sergii Leshchenko
+ */
+@Singleton
+public class RemoveLocalProjectsFolderOnWorkspaceRemove
+    implements EventSubscriber<WorkspaceRemovedEvent> {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(RemoveLocalProjectsFolderOnWorkspaceRemove.class);
+
+  private final ExecutorService executor;
+  private final LocalProjectsFolderPathProvider workspaceFolderPathProvider;
+
+  @Inject(optional = true)
+  @Named("host.projects.root")
+  private String hostProjectsFolder;
+
+  @Inject
+  public RemoveLocalProjectsFolderOnWorkspaceRemove(
+      LocalProjectsFolderPathProvider workspaceFolderPathProvider) {
+    this.workspaceFolderPathProvider = workspaceFolderPathProvider;
+    executor =
+        Executors.newCachedThreadPool(
+            new ThreadFactoryBuilder()
+                .setNameFormat("RemoveLocalProjectsFolderOnWorkspaceRemove-%d")
+                .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+                .setDaemon(true)
+                .build());
+  }
+
+  @Inject
+  public void subscribe(EventService eventService) {
+    eventService.subscribe(this);
+  }
+
+  @Override
+  public void onEvent(WorkspaceRemovedEvent event) {
+    Workspace workspace = event.getWorkspace();
+    String workspaceName = workspace.getConfig().getName();
+
+    String projectSourcesPath;
+    try {
+      projectSourcesPath = workspaceFolderPathProvider.getPathByName(workspaceName);
+    } catch (IOException e) {
+      LOG.error(
+          "Failed to evaluate projects files path for workspace with id: '{}'. Cause: '{}'",
+          workspace.getId(),
+          e.getMessage());
+      return;
+    }
+
+    if (!projectSourcesPath.equals(hostProjectsFolder)) {
+      // Remove the workspace folder located in the Che instance docker container.
+      deleteRecursiveAsync(
+          workspace.getId(), new File(projectSourcesPath.replace(getenv("CHE_INSTANCE"), "")));
+    }
+  }
+
+  private void deleteRecursiveAsync(String workspaceId, File toRemove) {
+    executor.execute(
+        ThreadLocalPropagateContext.wrap(
+            () -> {
+              boolean success = deleteRecursive(toRemove);
+              if (!success) {
+                LOG.error(
+                    "Failed to remove projects files for workspace with id: '{}'.", workspaceId);
```

Comment:
Please add path to not removed folder too

Reply:
Makes sense. Thanks for remark